text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+Moreover, custom errors from constructor parameter validity checks can be emitted during validation when using extension functions, but not when using Cedar functions.
+
+Second, there is no special parsing for Cedar functions. With Cedar's built-in `decimal`, you can write `decimal("123.12")` which more directly conveys the number being represented than does `mydecimal(123,12)`.
+
+Of course, these drawbacks do not necessarily speak against Cedar functions generally, but suggest that for suitably general use-cases (like decimal numbers!), an extension function might be warranted instead.
+
+### Readability: Policies are no longer standalone
+A policy can no longer be read by itself, it has to be read in the context of all function definitions it uses.
+Policies that use a large number of functions may be hard to read.
+
+## Alternatives
+
+### Type annotations
+We could require/allow Cedar functions to have type annotations, taking type definitions from either the schema or allowing them to be inline.
+
+Example:
+
+Schema:
+```
+type SemVer = { major : Long, minor : Long, patch : Long };
+```
+Policy:
+```
+function semver(major : Long, minor : Long, patch : Long) -> Semver {
+  { major : major, minor : minor, patch : patch }
+};
+```
+
+This would allow functions to typechecked in absence of a use-site, and allow for easier user specification of intent.
+It would also probably result in clearer type checker error messages.
+
+This introduces the following questions:
+
+1. Do we allow `type` declaration in policies, or just in schemas?
+2. Are type annotations enforced dynamically à la Contracts, or are they just ignored at runtime?
+  1. If they are dynamically enforced, that implies access to the schema to unfold type definitions. It also may introduce redundant type checking.
+3. Are type annotations required or optional?
+4. Will we have types to support generics, i.e., polymorphism?
+
+
+Leaving them out for this RFC only precludes only one future design decision, making type annotations required.
+This decision feels unlikely, as we want Cedar to be useful without using a schema/the validator.
+Adding _optional_ type annotations in the future is backwards compatible, but mandating type annotations may not be, due to the use of polymorphic functions.
+
+### Let functions call other functions
+As long as cycles are forbidden and functions as arguments are disallowed, we could allow functions to call other functions without sacrificing termination.
+However, the potential complexity explosion is high, and it's backwards compatible to add this later.
+
+<!-- ### Make parse errors runtime errors
+Make (any subset of) the following errors runtime errors instead of parse errors: 
+
+1. Application of non-function
+2. Non-application of function
+3. Function application with incorrect arity -->
+
+### Naming
+Should these really be called `function`s? They are actually `macro`s.
```

Comment:
I'm not sure I agree that these are macros, in the current design.  They are pure functions.

If we want to avoid associations with other languages' functions or macros, we could also pick a third term, like `snippet`, `predicate`, `def`, ....

Reply:
Fine with `snippet` or `def`

Reply:
I would claim these are best _communicated_ as macros, for two reasons:

1. Most readers are used to CBV languages. So they inherently assume (having not actually read the docs) that for _functions_ when you write `f(1+2,principal.id like "foo")` then you will evaluate `1+2` and then `principal.id like "foo"`, and then call the function with the results. They will not imagine that inlining will happen, and that short-circuiting and whatnot can change the results.
2. Readers are familiar with macros in C, and perhaps other languages. For _macros_ they know that when you write `f(1+2,principal.id like "foo")` you are substituting full expressions `1+2` and `principal.id like "foo"` into the body, you are not evaluating the them first. 

So I claim that by calling them macros we help users naturally go to the right semantics in their heads. Yes, it's true that we are really (or equivalently) providing is call-by-name functions, but calling them macros is more illuminating, IMO.

In sum: I propose changing the whole RFC to be "Cedar Function Macros" and then use "macro" for short, throughout.

Reply:
Writing up the function vs macro discussion in the RFC...

Reply:
I also prefer `macro`. To me `function` suggests that I should be able to nest calls, whereas I can easily accept that a macro shouldn't call another macro. I'm not a fan of `predicate` since that suggests that the resulting expression must be boolean-typed. No preference on `snippet`, except that I haven't seen it used before so it feels non-standard.

Also, not a fan of "Cedar Function Macros". I found this terminology confusing on my initial read through -- just pick function or macro.

Reply:
I'm happy with `macro`. Maybe `inline function` would be another option (though it still may suggest CBV to some readers)?